File: src/com/mingliqiye/utils/collection/Maps.java
package com.mingliqiye.utils.collection;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Maps工具类提供了一系列创建Map实现的便捷方法。
 *
 * @author MingLiPro
 */
public class Maps {

    /**
     * 创建一个空的HashMap实例。
     *
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 新创建的空HashMap实例
     */
    public static <K, V> Map<K, V> newHashMap() {
        return new HashMap<>();
    }

    /**
     * 创建一个指定初始容量的空HashMap实例。
     *
     * @param size 初始容量大小
     * @param <K>  Map键的类型
     * @param <V>  Map值的类型
     * @return 指定初始容量的空HashMap实例
     */
    public static <K, V> Map<K, V> newHashMap(int size) {
        return new HashMap<>(size);
    }

    /**
     * 根据已有Map创建一个新的HashMap实例。
     *
     * @param map 要复制的Map
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 包含原Map所有元素的新HashMap实例
     */
    public static <K, V> Map<K, V> newHashMap(Map<K, V> map) {
        Map<K, V> newMap = newHashMap();
        newMap.putAll(map);
        return newMap;
    }

    /**
     * 创建一个空的LinkedHashMap实例。
     *
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 新创建的空LinkedHashMap实例
     */
    public static <K, V> Map<K, V> newLinkedHashMap() {
        return new LinkedHashMap<>();
    }

    /**
     * 创建一个指定初始容量的空LinkedHashMap实例。
     *
     * @param size 初始容量大小
     * @param <K>  Map键的类型
     * @param <V>  Map值的类型
     * @return 指定初始容量的空LinkedHashMap实例
     */
    public static <K, V> Map<K, V> newLinkedHashMap(int size) {
        return new LinkedHashMap<>(size);
    }

    /**
     * 根据已有Map创建一个新的LinkedHashMap实例。
     *
     * @param map 要复制的Map
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 包含原Map所有元素的新LinkedHashMap实例
     */
    public static <K, V> Map<K, V> newLinkedHashMap(Map<K, V> map) {
        Map<K, V> newMap = newLinkedHashMap();
        newMap.putAll(map);
        return newMap;
    }

    /**
     * 创建一个空的TreeMap实例。
     *
     * @param <K> Map键的类型，必须实现Comparable接口
     * @param <V> Map值的类型
     * @return 新创建的空TreeMap实例
     */
    public static <K extends Comparable<K>, V> Map<K, V> newTreeMap() {
        return new TreeMap<>();
    }

    /**
     * 根据已有Map创建一个新的TreeMap实例。
     *
     * @param map 要复制的Map
     * @param <K> Map键的类型，必须实现Comparable接口
     * @param <V> Map值的类型
     * @return 包含原Map所有元素的新TreeMap实例
     */
    public static <K extends Comparable<K>, V> Map<K, V> newTreeMap(
        Map<K, V> map
    ) {
        Map<K, V> newMap = newTreeMap();
        newMap.putAll(map);
        return newMap;
    }

    /**
     * 创建一个空的Hashtable实例。
     *
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 新创建的空Hashtable实例
     */
    public static <K, V> Map<K, V> newHashtable() {
        return new Hashtable<>();
    }

    /**
     * 创建一个指定初始容量的空Hashtable实例。
     *
     * @param size 初始容量大小
     * @param <K>  Map键的类型
     * @param <V>  Map值的类型
     * @return 指定初始容量的空Hashtable实例
     */
    public static <K, V> Map<K, V> newHashtable(int size) {
        return new Hashtable<>(size);
    }

    /**
     * 根据已有Map创建一个新的Hashtable实例。
     *
     * @param map 要复制的Map
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 包含原Map所有元素的新Hashtable实例
     */
    public static <K, V> Map<K, V> newHashtable(Map<K, V> map) {
        Map<K, V> newMap = newHashtable();
        newMap.putAll(map);
        return newMap;
    }

    /**
     * 创建一个空的ConcurrentHashMap实例。
     *
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 新创建的空ConcurrentHashMap实例
     */
    public static <K, V> Map<K, V> newConcurrentHashMap() {
        return new ConcurrentHashMap<>();
    }

    /**
     * 创建一个指定初始容量的空ConcurrentHashMap实例。
     *
     * @param size 初始容量大小
     * @param <K>  Map键的类型
     * @param <V>  Map值的类型
     * @return 指定初始容量的空ConcurrentHashMap实例
     */
    public static <K, V> Map<K, V> newConcurrentHashMap(int size) {
        return new ConcurrentHashMap<>(size);
    }

    /**
     * 根据已有Map创建一个新的ConcurrentHashMap实例。
     *
     * @param map 要复制的Map
     * @param <K> Map键的类型
     * @param <V> Map值的类型
     * @return 包含原Map所有元素的新ConcurrentHashMap实例
     */
    public static <K, V> Map<K, V> newConcurrentHashMap(Map<K, V> map) {
        Map<K, V> newMap = newConcurrentHashMap();
        newMap.putAll(map);
        return newMap;
    }
}

File: src/main/kotlin/com/mingliqiye/utils/map/Maps.kt
/*
 * Copyright 2025 mingliqiye
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ProjectName mingli-utils
 * ModuleName mingli-utils.main
 * CurrentFile Maps.kt
 * LastUpdate 2025-09-15 08:33:24
 * UpdateUser MingLiPro
 */

@file:JvmName("Maps")

package com.mingliqiye.utils.map

import java.util.*
import java.util.concurrent.ConcurrentHashMap


fun <K, V> newHashMap(vararg elements: Pair<K, V>): HashMap<K, V> {
    val map = HashMap<K, V>()
    for (element in elements) {
        map[element.first] = element.second
    }
    return map
}

fun <K, V> newHashMap(size: Int): HashMap<K, V> {
    return HashMap(size)
}

fun <K, V> newHashMap(vararg elements: Map.Entry<K, V>): HashMap<K, V> {
    val map = HashMap<K, V>()
    for (element in elements) {
        map[element.key] = element.value
    }
    return map
}

fun <K, V> newConcurrentHashMap(vararg elements: Map.Entry<K, V>): ConcurrentHashMap<K, V> {
    val map = newConcurrentHashMap<K, V>()
    for (element in elements) {
        map[element.key] = element.value
    }
    return map
}

fun <K, V> newLinkedHashMap(vararg elements: Map.Entry<K, V>): LinkedHashMap<K, V> {
    val map = newLinkedHashMap<K, V>()
    for (element in elements) {
        map[element.key] = element.value
    }
    return map
}

fun <K, V> newHashtable(vararg elements: Map.Entry<K, V>): Hashtable<K, V> {
    val map = newHashtable<K, V>()
    for (element in elements) {
        map[element.key] = element.value
    }
    return map
}

fun <K : Comparable<K>, V> newTreeMap(vararg elements: Map.Entry<K, V>): TreeMap<K, V> {
    val map = newTreeMap<K, V>()
    for (element in elements) {
        map[element.key] = element.value
    }
    return map
}

inline fun <reified K, reified V> Map<K, V>.newHashMap(): HashMap<K, V> {
    return HashMap(this)
}

fun <K, V> newLinkedHashMap(): LinkedHashMap<K, V> {
    return LinkedHashMap()
}

fun <K, V> newLinkedHashMap(size: Int): LinkedHashMap<K, V> {
    return LinkedHashMap(size)
}

fun <K, V> newLinkedHashMap(vararg elements: Pair<K, V>): LinkedHashMap<K, V> {
    val map = newLinkedHashMap<K, V>()
    for (element in elements) {
        map[element.first] = element.second
    }
    return map
}

inline fun <reified K, reified V> Map<K, V>.newLinkedHashMap(): LinkedHashMap<K, V> {
    return LinkedHashMap(this)
}

fun <K : Comparable<K>, V> newTreeMap(): TreeMap<K, V> {
    return TreeMap()
}

fun <K : Comparable<K>, V> newTreeMap(vararg elements: Pair<K, V>): TreeMap<K, V> {
    val map = newTreeMap<K, V>()
    for (element in elements) {
        map[element.first] = element.second
    }
    return map
}

inline fun <reified K : Comparable<K>, reified V> Map<K, V>.newTreeMap(): TreeMap<K, V> {
    return TreeMap(this)
}

fun <K, V> newHashtable(): Hashtable<K, V> {
    return Hashtable()
}

fun <K, V> newHashtable(size: Int): Hashtable<K, V> {
    return Hashtable(size)
}

fun <K, V> newHashtable(vararg elements: Pair<K, V>): Hashtable<K, V> {
    val map = newHashtable<K, V>()
    for (element in elements) {
        map[element.first] = element.second
    }
    return map
}

inline fun <reified K, reified V> Map<K, V>.newHashtable(): Hashtable<K, V> {
    val hashtable = Hashtable<K, V>()
    hashtable.putAll(this)
    return hashtable
}

fun <K, V> newConcurrentHashMap(): ConcurrentHashMap<K, V> {
    return ConcurrentHashMap()
}

fun <K, V> newConcurrentHashMap(size: Int): ConcurrentHashMap<K, V> {
    return ConcurrentHashMap(size)
}

fun <K, V> newConcurrentHashMap(vararg elements: Pair<K, V>): ConcurrentHashMap<K, V> {
    val map = newConcurrentHashMap<K, V>()
    for (element in elements) {
        map[element.first] = element.second
    }
    return map
}

inline fun <reified K, reified V> Map<K, V>.newConcurrentHashMap(): ConcurrentHashMap<K, V> {
    return ConcurrentHashMap(this)
}

fun <K, V> ofEntries(entries: List<MutableMap.MutableEntry<K, V>>): Map<K, V> {
    return entries.associate { it.key to it.value }
}
